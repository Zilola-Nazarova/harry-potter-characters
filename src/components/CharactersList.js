import { useState } from 'react';
import { Link } from 'react-router-dom';
import { useSelector } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import { BsArrowRightCircle } from 'react-icons/bs';
import styles from '../styles/CharactersList.module.css';

const CharactersList = () => {
  const { gpCharacters, isLoading, error } = useSelector((store) => store.gpCharacters);
  const [inputText, setInputText] = useState('');
  const filtered = gpCharacters.filter((person) => person.name.toLowerCase().includes(inputText));
  const inputHandler = (e) => {
    const lowerCase = e.target.value.toLowerCase();
    setInputText(lowerCase);
  };

  return (
    <div className={styles.home}>
      <div className={styles.search}>
        <h1>Harry Potter Characters</h1>
        <input
          placeholder="search..."
          onChange={inputHandler}
        />
      </div>
      {(isLoading) && <p className="loading">The list of characters is loading...</p>}

      {(error) && (
        <p className="error">
          Something went wrong!
          <br />
          { error }
        </p>
      )}

      {(filtered) && (
        <>
          {(filtered.length === 0) && (
            <p className="error">
              Something went wrong!
              <br />
              No characters found.
            </p>
          )}
          <ul className={styles.list}>
            {filtered.map((person) => (
              <li
                key={uuidv4()}
                className={styles.card}
              >
                <Link to={`details/${person.name}`}>
                  <BsArrowRightCircle className={styles.icon} />
                  <div className={styles.portrait}>
                    {(person.image)
                      ? <img alt="Character portrait" src={person.image} />
                      : <p>no image available</p>}
                  </div>
                  <div className={styles.info}>
                    <h2 className={styles.name}>{person.name}</h2>
                    <h3 className={styles.house}>{person.house}</h3>
                    <p className={styles.alive}>{(JSON.stringify(person.alive)) ? 'alive' : 'not alive'}</p>
                  </div>
                </Link>
              </li>
            ))}
          </ul>
        </>
      )}
    </div>
  );
};

export default CharactersList;
