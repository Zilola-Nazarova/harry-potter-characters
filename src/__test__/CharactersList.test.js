import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import CharactersList from '../components/CharactersList';

const mock = [
  {
    id: '9e3f7ce4-b9a7-4244-b709-dae5c1f1d4a8',
    name: 'Harry Potter',
    alternate_names: ['The Boy Who Lived', 'The Chosen One'],
    species: 'human',
    gender: 'male',
    house: 'Gryffindor',
    dateOfBirth: '31-07-1980',
    yearOfBirth: 1980,
    wizard: true,
    ancestry: 'half-blood',
    eyeColour: 'green',
    hairColour: 'black',
    wand: {
      wood: 'holly',
      core: 'phoenix feather',
      length: 11,
    },
    patronus: 'stag',
    hogwartsStudent: true,
    hogwartsStaff: false,
    actor: 'Daniel Radcliffe',
    alternate_actors: [],
    alive: true,
    image: 'https://ik.imagekit.io/hpapi/harry.jpg',
  },
  {
    id: '4c7e6819-a91a-45b2-a454-f931e4a7cce3',
    name: 'Hermione Granger',
    alternate_names: [],
    species: 'human',
    gender: 'female',
    house: 'Gryffindor',
    dateOfBirth: '19-09-1979',
    yearOfBirth: 1979,
    wizard: true,
    ancestry: 'muggleborn',
    eyeColour: 'brown',
    hairColour: 'brown',
    wand: {
      wood: 'vine',
      core: 'dragon heartstring',
      length: 10.75,
    },
    patronus: 'otter',
    hogwartsStudent: true,
    hogwartsStaff: false,
    actor: 'Emma Watson',
    alternate_actors: [],
    alive: true,
    image: 'https://ik.imagekit.io/hpapi/hermione.jpeg',
  },
];
const reducer = (
  state = {
    gpCharacters: { gpCharacters: mock, isLoading: false, error: undefined },
  },
) => state;

const store = configureStore({ reducer });

test('Check if the Root component is rendered correctly', () => {
  const tree = render(
    <Provider store={store}>
      <MemoryRouter>
        <CharactersList />
      </MemoryRouter>
    </Provider>,
  );

  expect(tree).toMatchSnapshot();
  expect(screen.getByText('Hermione Granger')).toBeInTheDocument();
});
