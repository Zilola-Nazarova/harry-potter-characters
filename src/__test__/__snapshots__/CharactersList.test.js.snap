// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check if the Root component is rendered correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="search"
      >
        <h1>
          Harry Potter Characters
        </h1>
        <input
          placeholder="search..."
        />
      </div>
      <ul
        class="list"
      >
        <li
          class="card"
        >
          <a
            href="/details/Harry Potter"
          >
            <svg
              class="icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
                fill-rule="evenodd"
              />
            </svg>
            <div
              class="portrait"
            >
              <img
                alt="Character portrait"
                src="https://ik.imagekit.io/hpapi/harry.jpg"
              />
            </div>
            <div
              class="info"
            >
              <h2
                class="name"
              >
                Harry Potter
              </h2>
              <h3
                class="house"
              >
                Gryffindor
              </h3>
              <p
                class="alive"
              >
                alive
              </p>
            </div>
          </a>
        </li>
        <li
          class="card"
        >
          <a
            href="/details/Hermione Granger"
          >
            <svg
              class="icon"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
                fill-rule="evenodd"
              />
            </svg>
            <div
              class="portrait"
            >
              <img
                alt="Character portrait"
                src="https://ik.imagekit.io/hpapi/hermione.jpeg"
              />
            </div>
            <div
              class="info"
            >
              <h2
                class="name"
              >
                Hermione Granger
              </h2>
              <h3
                class="house"
              >
                Gryffindor
              </h3>
              <p
                class="alive"
              >
                alive
              </p>
            </div>
          </a>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <div
      class="search"
    >
      <h1>
        Harry Potter Characters
      </h1>
      <input
        placeholder="search..."
      />
    </div>
    <ul
      class="list"
    >
      <li
        class="card"
      >
        <a
          href="/details/Harry Potter"
        >
          <svg
            class="icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
              fill-rule="evenodd"
            />
          </svg>
          <div
            class="portrait"
          >
            <img
              alt="Character portrait"
              src="https://ik.imagekit.io/hpapi/harry.jpg"
            />
          </div>
          <div
            class="info"
          >
            <h2
              class="name"
            >
              Harry Potter
            </h2>
            <h3
              class="house"
            >
              Gryffindor
            </h3>
            <p
              class="alive"
            >
              alive
            </p>
          </div>
        </a>
      </li>
      <li
        class="card"
      >
        <a
          href="/details/Hermione Granger"
        >
          <svg
            class="icon"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1 8a7 7 0 1 0 14 0A7 7 0 0 0 1 8zm15 0A8 8 0 1 1 0 8a8 8 0 0 1 16 0zM4.5 7.5a.5.5 0 0 0 0 1h5.793l-2.147 2.146a.5.5 0 0 0 .708.708l3-3a.5.5 0 0 0 0-.708l-3-3a.5.5 0 1 0-.708.708L10.293 7.5H4.5z"
              fill-rule="evenodd"
            />
          </svg>
          <div
            class="portrait"
          >
            <img
              alt="Character portrait"
              src="https://ik.imagekit.io/hpapi/hermione.jpeg"
            />
          </div>
          <div
            class="info"
          >
            <h2
              class="name"
            >
              Hermione Granger
            </h2>
            <h3
              class="house"
            >
              Gryffindor
            </h3>
            <p
              class="alive"
            >
              alive
            </p>
          </div>
        </a>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
